packages/opencensus-web-instrumentation-zone/src/perf-resource-timing-selector.ts

```diff
@@ -0,0 +1,169 @@
+ * name (it should match the XHR URL), additionally, the start/end timings of
+ * every performance entry should fit within the span start/end timings.
+ * These filtered performance resource entries are considered as possible
+ * entries associated to the xhr.
+ * Those are possible because there might be more than two entries that pass the
+ * filter.
+ * @param xhrUrl
+ * @param span
+ */
+export function getPerfResourceEntries(
+  xhrUrl: string,
+  span: Span
+): PerformanceResourceTiming[] {
+  return performance
+    .getEntriesByType('resource')
+    .filter(entry =>
+      isPerfEntryPartOfXhr(entry as PerformanceResourceTiming, xhrUrl, span)
+    ) as PerformanceResourceTiming[];
+}
+
+/**
+ * Second step for the 'Performance resource timings selector algorithm'.
+ * As the XHR could cause a CORS pre-flight request, we have to look for
+ * possible performance entries either containing cors preflight timings or not.
+ * A possible entry with cors data is when a resource timing entry does not
+ * overlap timings with other resource timing entry. Also, a possible entry
+ * without cors resource timing is when that resource timing entry is not
+ * 'paired' with any other entry.
+ * Thus, for this step traverse the array of resource entries and for every
+ * entry check if it is a possible performance resource entry.
+ * @param perfEntries
+ */
+export function getPossiblePerfResourceEntries(
+  perfEntries: PerformanceResourceTiming[]
+): XhrPerformanceResourceTiming[] {
+  const possiblePerfEntries = new Array<XhrPerformanceResourceTiming>();
+  const pairedEntries = new Set<PerformanceResourceTiming>();
+  let entryI: PerformanceResourceTiming;
+  let entryJ: PerformanceResourceTiming;
+  // As this part of the algorithm traverses the array twice, although,
+  // this array is not large as the performance resource entries buffer is
+  // cleared when there are no more running XHRs.
+  for (let i = 0; i < perfEntries.length; i++) {
+    entryI = perfEntries[i];
+    // Compare every performance entry with its consecutive perfomance entries.
+    // That way to avoid comparing twice the entries.
+    for (let j = i + 1; j < perfEntries.length; j++) {
+      entryJ = perfEntries[j];
+      if (!overlappingPerfResourceTimings(entryI, entryJ)) {
+        // As the entries are not overlapping, that means those timings
+        // are possible perfomance timings related to the XHR.
+        possiblePerfEntries.push({
+          corsPreFlightRequest: entryI,
+          mainRequest: entryJ,
+        });
+        pairedEntries.add(entryI);
+        pairedEntries.add(entryJ);
+      }
+    }
+    // If the entry couldn't be paired with any other resource timing,
+    // add it as a possible resource timing without cors preflight data.
+    // This is possible because this entry might be better than the other
+    // possible entries.
+    if (!pairedEntries.has(entryI)) {
```

Comment:
Can we remove this `if` statement and still have the algorithm work? I'm not 100% sure yet that it is always correct to have the if, and anyway removing it could make it simpler - I'm open if I'm wrong on this!

Reply:
My decision was to remove this part and do not use this Set anymore, this to remove complexity on the algorithm. However it still is working as expected.